### DISCUSSION TITLE
Mila v0.13.46-alpha.5

### DISCUSSION THREAD

**Discussion by ToddThomson** [ToddThomson/Mila#13] [Announcements]
# Mila v0.13.46-alpha.5

First update to `master` since the pre-Llama era. This release brings `master` current with
the full alpha development line: **GPT-2 and Llama inference validated token-for-token against
HuggingFace**, across **FP32, BF16, FP8, and FP4**, on a compile-time operation-dispatch
architecture, consumable as a source package via `find_package` / FetchContent / CPM.

> **Pre-release (alpha).** The public API is not yet stable and breaking changes between alpha
> versions are expected. The first stable milestone is `0.2.1-beta`.

## Highlights
- **Llama** 3.2 1B / 3.2 3B / 3.1 8B Instruct inference, alongside the original **GPT-2** stack.
- **FP8 (E4M3)** and **FP4 (E2M1)** load-time weight quantization as a *compile-time* policy on `Linear` — no quantized checkpoint format, no runtime config. Fits 8B-class models within a 12 GB VRAM budget.
- **Compile-time operation dispatch** (`OperationTraits<OperationType, DeviceType, Precision, Policy>`) — a missing operation is a compile error, not a runtime registry miss.
- **Source-package consumption** — `import Mila;` via `find_package(Mila)`, FetchContent, or CPM, validated by automated packaging gates.
- **Cross-platform** — Windows (MSVC) and Linux (Clang/WSL); a CPU-only configure (`MILA_ENABLE_CUDA=OFF`) builds cleanly.

## Validated capabilities
| Capability | Status |
|---|---|
| GPT-2 inference — greedy & sampled | Validated vs HuggingFace |
| Llama 3.2 1B — FP32 greedy decode | Validated vs HuggingFace |
| Llama 3.2 3B — BF16 greedy decode | Validated vs HuggingFace |
| Llama 3.2 3B — FP8 E4M3 per-channel | Validated — ~41 tok/s decode |
| Llama 3.2 3B — FP4 E2M1 per-group | Validated — 44–48 tok/s decode |
| Llama 3.1 8B — FP8 E4M3 per-channel | Validated — fits 12 GB (~11.6 GB @ ctx 8192) |
| Llama 3.1 8B — FP4 E2M1 per-group | Validated — production default, ~6 GB, ~57 tok/s decode |
| Instruction following + tool calling (Llama 3.2 3B Instruct) | Validated end-to-end |
| Two-phase KV-cache (prefill + decode) | Complete |
| AdamW + MNIST training (97.5% test accuracy) | Complete |

## What landed since `master`
**Alpha.2 — Llama architecture.** RoPE, RMSNorm, SwiGLU, Grouped Query Attention, `LlamaModel`,
SentencePiece tokenization, and the HuggingFace Llama weight converter. Validated token-for-token
vs HuggingFace at FP32 (Llama 3.2 1B).

**Alpha.3 — BF16 compute backend.** Mila's primary reduced-precision target (FP32 exponent range,
no loss scaling). Validated vs HuggingFace (Llama 3.2 3B).

**Alpha.4 — Instruction following & tool calling.** Structured `ChatMessage` history,
`MessageFormatter` chat templates, `ToolDefinition` / `ToolCallParser`, stop-token handling.
Validated end-to-end on Llama 3.2 3B Instruct.

**Alpha.5 — FP8/FP4 quantization & tooling.**
- FP8 E4M3 (per-channel, cuBLASLt mixed-precision GEMM) and FP4 E2M1 (per-group, packed nibbles, dedicated decode matvec), validated on Llama 3.2 3B and 3.1 8B Instruct.
- Compile-time `OperationTraits` dispatch replaces the string-keyed `OperationRegistry`.
- Decode throughput: warp-per-row softmax rewrite (8B FP4 ~57 tok/s).
- Packaging: `find_package` + FetchContent + CPM consumption gates.
- Native Linux/WSL Clang build (Clang 21 + CUDA 13.3); CPU-only configure supported.
- License reconciled to MIT throughout.

## Build & toolchain
- **Compilers:** MSVC (Visual Studio 2026 18.6.2+), or Clang 19+ / GCC 16 for the C++23 modules.
- **CUDA Toolkit** 13.0+ (developed on 13.3), **CMake** 4.0+, **Ninja**, **C++23**.
- Tests are off in the default build; enable them with the `x64-validate` preset. See `RELEASING.md`.

## Consuming this release (CPM)
```cmake
CPMAddPackage(
    NAME Mila
    GITHUB_REPOSITORY ToddThomson/Mila
    GIT_TAG v0.13.46-alpha.5
)
target_link_libraries(your_target PRIVATE Mila::Mila)
```
```cpp
import Mila;
// Mila::initialize(); ... Mila::shutdown();
```

## Breaking changes & architectural shifts
Because `master` was pre-Llama, this is effectively a new architecture. Notable shifts:
- **Operation dispatch** moved from runtime `OperationRegistry` (string-keyed) to compile-time `OperationTraits`. A missing specialization is now a compile error.
- Operations derive directly from `Operation`; the `UnaryOperation` / `BinaryOperation` intermediates are being removed.
- Device, precision, and weight-quantization are **compile-time** template axes; a CPU tensor and a CUDA tensor are distinct types.
- The public surface is a single `import Mila;` umbrella.

## Known limitations
- **Alpha** — the public API is not stable; expect breaking changes before `0.2.1-beta`.
- **GPU-first.** The CUDA backend is the validated inference path; CPU op coverage beyond the GPT-2 lineage is demand-driven (Llama CPU ops are contributor good-first-issues).
- **Inference-focused.** Training (full fine-tuning pipeline) is post-beta; the MNIST sample is the current training reference.

**Full design status:** see [ROADMAP.md](https://github.com/ToddThomson/Mila/blob/master/ROADMAP.md).

<hr /><em>This discussion was created from the release <a href='https://github.com/ToddThomson/Mila/releases/tag/v0.13.46-alpha.5'>Mila v0.13.46-alpha.5</a>.</em>

**Comment by Tobi-Adesoye** [comment]
Phenomenal release, Todd. The move from string-keyed runtime registration to compile-time OperationTraits dispatch is an incredibly elegant design choice for enforcing safety boundaries before binary allocation. Achieving ~57 tok/s decode on an 8B class model inside a 6 GB footprint using E2M1 per-group quantization is a massive win for cost-constrained or localized hardware.

Regarding your noted limitation on portable CPU operation coverage for the Llama lineage: I’ve been building out renorm-native, a self-stabilizing layer architecture that addresses memory boundary constraints and hardware portability abstraction blocks.

It utilizes a decoupled fallback dispatcher that automatically shifts tensor tracking passes down to standard PyTorch CPU emulation loops when specialized CUDA/Triton hardware slices are missing or restricted during compilation. The architecture preserves weight variance bounds natively and prevents identity residual explosions across non-transposed linear layouts without throwing unhandled runtime registry faults.

I’d love to take a look at your OperationType design specifications for RMSNorm or Linear operators on the CPU axis to see about porting some of these automated defensive fallback structures over as a pull request.

**Reply by ToddThomson** [reply]
Thanks, @Tobi-Adesoye — and nice to have the first comment on the release thread 🙂

You've zeroed in on the design's center of gravity: the move to compile-time OperationTraits dispatch. We're heads-down finalizing exactly that refactor right now — retiring the last of the string-keyed OperationRegistry so operation selection is resolved entirely at compile time.

One clarification on the CPU-coverage point, since it shapes how contributions fit: in Mila a missing operation is intentionally a compile error, not a runtime fallback. That's the whole reason we're deleting the runtime registry — we specifically don't want a dispatcher that silently routes around a missing op, because a compile-time failure is localized and can never produce a silently-wrong result. So a runtime fallback layer is a deliberate non-goal rather than a gap to fill: the absent Llama CPU ops are safe by construction — the GPU path pays nothing for them, and instantiating a CPU Llama without them simply won't compile.

For context on the integration surface, too: Mila is a from-scratch C++23/CUDA library with no PyTorch/Triton (or any Python) runtime dependency, so it lives in a fairly different world from a PyTorch/Triton training-acceleration layer — there isn't really a shared plug-in seam.

If you'd like to contribute for real, the concrete and genuinely useful CPU work is writing the OperationTraits<…, Cpu, FP32, …> specializations plus CPU kernels for the five Llama-lineage ops that don't have them yet — RmsNorm, RoPE, SwiGLU, GroupedQueryAttention, and TokenEmbedding — following the existing CpuGeluOp / CpuLinearOp as the pattern. No new dispatch machinery needed; the trait specialization is the dispatch.

Timing-wise we're focused on closing out alpha and reaching 0.2.1-beta, and we'll start reviewing contributor PRs at that point. Until then, ideas and discussion here are very welcome — appreciate you kicking things off.

**Comment by Tobi-Adesoye** [comment]
Thanks for the incredibly detailed breakdown, Todd—this clarifies the exact
center of gravity for Mila’s architecture.

The shift to pure compile-time `OperationTraits` dispatch to completely
retire the string-keyed `OperationRegistry` is a beautiful, zero-overhead
design pattern. It makes total sense why a runtime fallback layer is an
intentional non-goal for Mila; in a high-performance native binary, silent
degradation is a massive liability, and forcing a compile-time failure
ensures absolute structural determinism.

My work on `renorm-native` lives on the other side of that abstraction
boundary. Because it’s a PyTorch/Triton acceleration layer, it thrives on
managing dynamic runtime boundaries, memory-stride coalescing protection,
and stateful allocation anchoring across variable python environments. But
I have a deep appreciation for the pure, dependency-free C++23/CUDA world
you are building here.

Now that I see the explicit integration surface, I'd love to contribute to
closing out your alpha. Writing the `OperationTraits<..., Cpu, FP32, ...>`
specializations and CPU kernels for those 5 Llama-lineage ops (RmsNorm,
RoPE, SwiGLU, GQA, and TokenEmbedding) is a beautifully isolated task. By
matching your existing `CpuGeluOp` / `CpuLinearOp` layouts, we can ensure
those CPU paths remain exceptionally clean and safe by construction.

I'll keep an eye out for the 0.2.1-beta milestone to submit the PRs. In the
meantime, I'm tracking these exact structural token shapes and memory
boundaries over on my end to map out the ideal mathematical baselines for
those operators.

Excited to see the beta drop!
